Implement markdown cell attachments. Allow drag’n’drop of images into…

julienr opened the pull request:
This adds a way for users to insert inline images in markdown cells using drag and drop

TODO:
- [x] Add ability to copy-paste images
- [x] Add a menu item to insert images
- [x] Test the "Insert Image" menu item. Seems like casperjs doesn't support drag'n'drop/paste testing for now.
- [x] Add Edit menu items to copy/paste cell attachments
  - Edit : Cut Cell Attachments
  - Edit : Copy Cell Attachments
  - Edit : Paste Cell Attachments
- [x] "Garbage collect" attachments that are not referenced from markdown ? For example when the notebook is saved.
- [x] There is a large delay when pasting large images from the clipboard until they appear in the markdown. Should show some visual indicator that something is going on
- [x] Large attachments warning ?

Ideas :
- [x] ~~Could add a ipython magic so you can access attachments from python code. Could be nice to directly attach some data to the notebook for small datasets.~~ ([won't happen](https://github.com/jupyter/nbformat/pull/21))
- Could add auto-completion for attachments when editing ?

This is an experimental implementation of what was discussed in #613 by @ellisonbg and @stevengj .
When a user drag and drops an image into a markdown cell, the image data is stored in the cell metadata and can be used as the source for an image by using a special 'nbdata:filename' syntax.

Here are some things to discuss to move forward :
- It should be possible to insert inline images without drag'n'drop. Add a menu item ?
- There should be a UI so the user can easily see which cells have 'attachments' and a way to manage them.

Screenshots
-----------------
Just after the drop of an image :
<img width="962" alt="1_source" src="https://cloud.githubusercontent.com/assets/506602/10591903/14d49614-76b9-11e5-90a9-0246972bb274.png">

The metadata :
<img width="689" alt="2_metadata" src="https://cloud.githubusercontent.com/assets/506602/10591919/26d009ac-76b9-11e5-9ff4-3a3f475b6970.png">

The rendered cell
<img width="819" alt="3_rendered" src="https://cloud.githubusercontent.com/assets/506602/10591925/2c245d86-76b9-11e5-9829-e0bf966659c8.png">

stevengj commented:
Great, thanks so much for working on this!

* Shouldn't it use [markdown image syntax](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#images) rather than `<img ../>`?

* Maybe it should use the notebook's generic "mime bundle" format, i.e. `"test.png" : { "image/png" : ["base64-encoded-png-data"] }`

* Would be good to support pasted images as well as drag-and-drop, and a menu item to insert an image from a file-picker menu.

jdfreder commented:
This is awesome!  Great work!

> Shouldn't it use markdown image syntax rather than <img ../>?

+1

> Would be good to support pasted images as well as drag-and-drop, and a menu item to insert an image from a file-picker menu.

+2

> Maybe it should use the notebook's generic "mime bundle" format, i.e. "test.png" : { "image/png" : ["base64-encoded-png-data"] }

It's not clear to me how this would work- where would the mime bundle be stored in the document and how would it be referenced inline in the markdown?

stevengj commented:
@jdfreder, the mime bundle would be stored in the cell data, exactly as above.  i.e.
```
{
   "collapsed": true,
    ....
    "attachments": {
        "test.png" : { "image/png": [...base64-encoded-data...] }
    }
}
```
The point is, the notebook format already has defined a way to store MIME data; you should re-use it here rather than inventing a new method.

(MIME bundles also open the door to attachments that are stored in multiple formats, which might be useful in the future.)

jdfreder commented:
Ah I didn't realize this was already being stored in the notebook metadata, I thought he was inlining the b64 content in the markdown.  I don't think we should be storing this data in notebook metadata, which is meant for users and extensions.  Instead I think it should have it's own key in [nbformat](https://github.com/jupyter/nbformat).

stevengj commented:
@jdfreder, it's not being stored in the notebook metadata, it's being stored as an extra key/value pair in the markdown *cell* data.   See the discussion in #613 — neither storing the data inline in the markdown nor storing it in a separate file are really acceptable long-term solutions, and @ellisonbg does not want to add a new cell type.

jdfreder commented:
> it's being stored as an extra key/value pair in the markdown cell data

But it's not, it's being stored in `that.metadata.attachments ` which is metadata - yes you're right, not notebook level.  All I'm trying to say that I think this needs another PR to https://github.com/jupyter/nbformat to change nbformat.  *Not to add another cell type*, but to add an `attachments` key to the cell.  

Metadata is not something we want to use inside Jupyter/notebook in an attempt to avoid making changes to nbformat.  If we are going to add keys to the notebook file, they need to be documented, and nbformat is how we document it.

julienr commented:
Thanks for the comments !

I added [a nbformat PR](https://github.com/jupyter/nbformat/pull/21) for a new 'attachments' cell property. 

I also added a todo list to my first comment on this issue to track progress.

ellisonbg commented:
A few points:

* +1 on using mime bundles for this
* +1 on adding the attachment field to the nbformat rather than metadata
* +1 on getting this to work both in img tags and markdown image syntax.
While the markdown image syntax is nice, I often need more control and use
igm tags.

On Tue, Oct 20, 2015 at 5:21 AM, Julien Rebetez <notifications@github.com>
wrote:

> Thanks for the comments !
>
> I added a nbformat PR <https://github.com/jupyter/nbformat/pull/21> for a
> new 'attachments' cell property.
>
> I also added a todo list to my first comment on this issue to track
> progress.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/621#issuecomment-149489680>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

stevengj commented:
@jdfreder, I see.  Yes, I completely agree; `attachments` should be a documented cell field.

stevengj commented on notebook/static/notebook/js/textcell.js:
```diff
@@ -259,6 +259,56 @@ define([
+        // cell, we do the following :
+        // - We insert the base64-encoded image into the cell attachments
+        //   directory, keyed by the filename.
+        // - We insert an img tag with a 'nbdata' src that refers to the
```
now an 'attachment:filename' source

stevengj commented on notebook/static/notebook/js/textcell.js:
```diff
@@ -259,6 +259,56 @@ define([
+        // Inline images insertion. When a user drops an image in a markdown
+        // cell, we do the following :
+        // - We insert the base64-encoded image into the cell attachments
+        //   directory, keyed by the filename.
```
directory -> dictionary?

julienr commented:
As suggested by @stevengj, I think it would be great to have an "insert image" menu item. This would be active only when the user is editing a markdown cell. I am not sure where to put it as it seems the current menu items apply to all cells and do not require one to be in edit mode. Do you have UI guidelines for that ?

ellisonbg commented:
I would find a suitable place in the "Edit" menu.

On Wed, Oct 21, 2015 at 3:09 AM, Julien Rebetez <notifications@github.com>
wrote:

> As suggested by @stevengj <https://github.com/stevengj>, I think it would
> be great to have an "insert image" menu item. This would be active only
> when the user is editing a markdown cell. I am not sure where to put it as
> it seems the current menu items apply to all cells and do not require one
> to be in edit mode. Do you have UI guidelines for that ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/621#issuecomment-149800587>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

minrk commented:
This wouldn't need any custom markdown processing at all, nor any change to the notebook format if drag&drop onto a markdown cell uploaded the file via contents API and then wrote a normal, relative URL in the markdown.

minrk commented:
The more I think about it, the more I think this should be uploading to contents, rather than writing binary data to the notebook. Doing so has many advantages:

- faster notebook load
- smaller storage size
- use existing tools to interact with images/data
- better version control behavior
- avoid need to parse special URLs in nbconvert
- avoid custom markdown processing
- easier to make attachments available to kernel by *not* embedding them in the notebook

I can't think of any benefits to the data being in the JSON in particular, just making it easier to put images into the notebook - the storage details of putting it into the JSON don't seem specifically valuable, or am I missing something?

julienr commented:
@minrk The advantage of having attachments in the JSON is that you have a single .ipynb file to deal with, for example when sharing your notebook via email. And you cannot "break" your notebook by moving files around. 

That is one advantage against the many advantages of the upload approach you listed though.

minrk commented:
That's a good point, it is a trade-off. We just have to decide which side we want to fall on. I suspect, given the consistent pressure we get that we already put too much in the JSON, that making it easier to work with external files (e.g. completion of filenames in markdown URLs) will be the preferable to yet-more-files in JSON. The drag&drop UI shouldn't be significantly different between the two storage schemes, though.

stevengj commented:
@minrk, having a single file to deal with is the whole point here.   I don't think we need a UI to help with linking external files — that is easy already: you just drag and drop the file into the same directory and add an image link to markdown.

Whereas right now it is insanely hard to embed an image in the notebook file itself, and this makes it extremely unreliable to share notebooks with markdown images without breaking them.

Moreover, when someone drags-and-drops or pastes an image into a notebook, what they will expect is that it is stored in the notebook.  This is the behavior people are familiar with from other contexts like word processors.  It is especially true for copy/paste or drag-and-drop of images that don't come from a file (e.g. pasting/dragging a selection from an image-editor program or from another web page).

stevengj commented:
@minrk, note also that 5 out of the 7 items in your list of advantages of separate files would also apply to images produce by code executions, i.e. to `display_data` messages.   And yet I think everyone can agree that the disadvantage of having to carry around separate image files for every graphical output would vastly outweigh the advantages.   The same applies to images in Markdown cells.

(Right now, in my experience, people simply avoid using images in markdown cells as much as possible because of this inconvenience.  If you make it easier to embed images, I suspect that you'll start seeing notebooks filled with technical illustrations and other graphical aids.)

juhasch commented:
The drag&drop notebook extension in IPython-contrib actually already uploads images to the server using the contents service. My experience is that for small images it sometimes would be nice to have them in the notebook, however I think it is better to store most images separately. 
I don't think JSON files make a good file system. The same is true for any local data that is being processed by a notebook.

So the discussions should probably be where to put local data that belongs to a notebook. 
For example, I have a lot of notebooks containing code, markdown, images. plots, external data.
Therefore I have only one or a few notebooks in a directory, together with graphics and data.

stevengj commented:
@juhasch, uploading things to the server does not help one bit with the problem of sharing the notebook with someone who does not share the same servers.   And again, the same argument would apply to storing plots and other code-generated images in the notebook vs. separate files.

juhasch commented:
Yes, I understand what you want and I would sometimes find that capability nice, too. 
The question is if JSON is a good format to be used as a container to store binary data, and  what kind of data  belongs into that container. I would argue we need a better container and not extend what is stored in the JSON file.

minrk commented:
> I think everyone can agree that the disadvantage of having to carry around separate image files for every graphical output would vastly outweigh the advantages

I certainly agree, but it's not everyone. We get regular complaints about putting output into the JSON.

> Right now, in my experience, people simply avoid using images in markdown cells as much as possible because of this inconvenience

And my experience is the opposite—since it's so convenient to use regular files in markdown and the behavior is so much better, I use relatively linked images in markdown all the time. For this reason, I thought that the convenience of upload was the only aspect this was meant to address, e.g. matching what GitHub image upload does.

> the same argument would apply to storing plots and other code-generated images in the notebook vs. separate files.

I don't agree, or at least I think the tradeoffs end up being different. Generated outputs and uploaded files are different in many ways. Outputs change on every execution of the notebook, and are reasonably discarded by users who do not want to track non-input content.

stevengj commented:
We're already using JSON as a container to store binary data.   I just want to be able to paste a few kB more in when I need a simple figure, without losing the self-contained nature of the file.

@minrk, I agree that the tradeoffs are even worse for code-generated images, but I wanted to point out that the arguments that you and @juhasch are making would largely apply to them as well.   That should give you pause — we have an existence proof that these arguments can be outweighed by the convenience of having a self-contained file.

ellisonbg commented:
I can see both sides here.

I see Min's point that almost everything is accomplished by uploading to
the server next to the notebook using the Content API.

I also see that we are already putting binary data in the notebook and
adding more won't causes too many problems.

From a usability perspective, I do think the main point is is allowing easy
uploading.

But the secondary point is to have a self contained file.


On Wed, Oct 21, 2015 at 12:42 PM, Steven G. Johnson <
notifications@github.com> wrote:

> We're already using JSON as a container to store binary data. I just want
> to be able to paste a few kB more in when I need a simple figure, without
> losing the self-contained nature of the file.
>
> @minrk <https://github.com/minrk>, I agree that the tradeoffs are even
> worse for code-generated images, but I wanted to point out that the
> arguments that you and @juhasch <https://github.com/juhasch> are making
> would largely apply to them as well. That should give you pause — we have
> an existence proof that these arguments can be outweighed by the
> convenience of having a self-contained file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/621#issuecomment-149956443>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

ellisonbg commented:
@jasongrout is probably in favor of embedding (next to me) - "I think the
call has already been made by how we handle output"

On Wed, Oct 21, 2015 at 12:59 PM, Brian Granger <ellisonbg@gmail.com> wrote:

> I can see both sides here.
>
> I see Min's point that almost everything is accomplished by uploading to
> the server next to the notebook using the Content API.
>
> I also see that we are already putting binary data in the notebook and
> adding more won't causes too many problems.
>
> From a usability perspective, I do think the main point is is allowing
> easy uploading.
>
> But the secondary point is to have a self contained file.
>
>
> On Wed, Oct 21, 2015 at 12:42 PM, Steven G. Johnson <
> notifications@github.com> wrote:
>
>> We're already using JSON as a container to store binary data. I just want
>> to be able to paste a few kB more in when I need a simple figure, without
>> losing the self-contained nature of the file.
>>
>> @minrk <https://github.com/minrk>, I agree that the tradeoffs are even
>> worse for code-generated images, but I wanted to point out that the
>> arguments that you and @juhasch <https://github.com/juhasch> are making
>> would largely apply to them as well. That should give you pause — we have
>> an existence proof that these arguments can be outweighed by the
>> convenience of having a self-contained file.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/jupyter/notebook/pull/621#issuecomment-149956443>.
>>
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu and ellisonbg@gmail.com
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

jasongrout commented:
I would say that there are usecases for both approaches, and right now only one approach is (easily) possible - the uploading and linking approach.  This makes it very easy to do the other approach too.

minrk commented:
If there is consensus that the trade offs are in favor of embedding in the JSON, that's fine by me. Sorry for the diversion.

julienr commented:
I would like to discuss how to handle attachments deletion. I see two possibilities :

1. We either automatically remove attachments that are not referenced in the markdown (at render time).
2. We let the user delete attachments manually using the "Edit Attachments" dialog (which should be improved).

Option 1 is easier for the user; he doesn't event need to know what an attachment is to use the feature. With this option, I am not sure an "Edit Attachments" dialog is still useful, although it could be used to view an attachments' size.

Any thoughts ?

stevengj commented:
Option 1 seems more intuitive to me.

stevengj commented:
What if you want to copy and paste an image from one markdown cell to another?

ellisonbg commented:
I would require explicit removal. It would be pretty easy to accidentally
render a cell when it temporarily didn't have a reference to the image. I
do this type of thing all the time when moving content around between
different cells.

To implement cut/copy/paste, I would create a subsection in the "Edit" menu
for these actions so folks don't have to monkey around with editing the
JSON metadata:

Edit : Cut Cell Attachments
Edit : Copy Cell Attachments
Edit : Paste Cell Attachments

On Thu, Oct 22, 2015 at 7:41 AM, Steven G. Johnson <notifications@github.com
> wrote:

> What if you want to copy and paste an image from one markdown cell to
> another?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/621#issuecomment-150194343>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

julienr commented:
@ellisonbg Good idea regarding the copy/paste.

A third option would be to require explicit removal, but show some kind of visual warning if you have an unreferenced attachment when rendering a markdown cell. So we get the intuitiveness of option 1 with the security (in term of you don't loose your content if you temporarily mess up your markdown) of option 2.

stevengj commented:
Any updates on this?

Carreau commented:
@parleur you said you wanted to have  a look at that yesterday.

julienr commented:
Sorry about the long delay, got busy with other things...
I'll find some time next week to rebase this against master and get it into a state suitable for a UI review.

parleur commented:
Hi. Wow it's an impressive work.

What can be done right now to push this feature forward?

Carreau commented:
> What can be done right now to push this feature forward?

Help @julienr to rebase (at least see the conflict there is), and the open points in firs message of this thread. 

likely but this can be separate: have nbconvert understand the attached images.

julienr commented:
I've started rebasing and will finish tomorrow. Then, I think some UI/UX review would be useful.

julienr commented:
I think I'm reaching a usable state with this PR. I'll try to add some JS tests for the attachment copy/paste feature.

I could use some feedback on the UI and the general usability of this. I'll also try to get some colleagues to test it next week. If you want to test it, you'll also need my [attachments branch](https://github.com/julienr/nbformat/tree/attachments) of nbformat.

There is one remaining bullet point which is to decide if users should be warned about unused attachments. For now, if you drag'n'drop a bunch of images in your markdown and then delete all the markdown, the attachments will stay unless you remove them with the attachment toolbar. I think it would be better to notify the user in some way. I see at least two solutions :

- Check for unused attachments when rendering a cell and display a visual markers on cells with unused attachments
- Check for unused attachments when saving the notebook and display a dialog warning the user
- Automatically garbage collect unused attachments when saving the notebook (I think this is dangerous because maybe the user has a good reason to have unused attachments)

What do you think ?

stevengj commented:
Warning about unused attachments only when saving the notebook seems good to me.  If the user has a good reason to have unused attachments, saving every time the cell is rendered seems too chatty.

parleur commented:
I have some remarks about unused cell attachment:

 - you told about good reason to have unused cell attachment? Except as a transitive state which is perfectly valid I cannot figure out why would a user save a notebook with it. At least as long as there is no way to know a cell have an attachment. 

 - Currently there are auto-save and manual save features. I think, checking, warning and or cleaning should only append when manual saved are performed. As embed arbitrary data in JSON is a hot point of debate, automatically remove unused cell attachment is a way to keep users safe from dirty uses (I can imagine a lot of it ).

I tested the UI, it seems quite clear for me and I havn't find any bug.

julienr commented:
@parleur I think you are right regarding unused cell attachments. If the user doesn't know the inner workings of cell attachments, there is no reason to except them to be saved if not used. I'll work on automatic removal on manual save.

Glad that the UI seems to work !

julienr commented:
I've added garbage collection on manual save. I think this is ready for a code review.

minrk commented on notebook/static/base/js/security.js:
```diff
@@ -31,6 +31,10 @@ define([
+        // TODO(julienr): This is a ugly hack to work around the fact that
+        // by default caja doesn't seem to like base64 src for inline markdown
+        // images. Not sure if we should do that here
+        ATTRIBS['img::src'] = 0;
```
Since this removes a security check, we need to make sure that there are no possible `image src:` attributes that can execute javascript before we can allow this.

julienr replied:
The good solution would be to only allow `data` uris (and disallow `javascript:` uri). Unfortunately, it seems that caja has hardcoded allowed uri schemes (and `data` is not included):

https://github.com/minrk/google-caja-bower/blob/master/html-sanitizer.js#L813

Should I just monkey patch `caja.html.ALLOWED_URI_SCHEMES` to allow data uris ?

There is an open issue on caja on this subject, but it doesn't seem very alive :
https://github.com/google/caja/issues/1558

minrk replied:
After a bit of reading, I [believe](https://www.owasp.org/index.php/Script_in_IMG_tags) this is not an issue on any supported browser. Can you make a note here that img:src is safe on browsers post ie6?

julienr commented:
I have some trouble understanding the failed travis tests. This seems to be related to the new `deprecated-require-paths.js` and I can reproduce the failure locally if I rebase my changes on master, but I am at a loss to understand what's causing the errors. Especially since I don't think I touched anything to do with requirejs...

For example in the following session, how can I debug what's causing the `Script error for "notebook"` ?
````
(jupydev)notebook|inline_images!=> python -m notebook.jstest  base/deprecated-require-paths.js

Test group: base/deprecated-require-paths.js
Running tests with notebook directory '/var/folders/x3/5y72x0994c7b32kr_kj8yg6w0000gn/T/tmphkDI2
y'
Test file: notebook/tests/base/deprecated-require-paths.js
Page Error
  Message:   Error: Script error for "notebook"
             http://requirejs.org/docs/errors.html#scripterror
  Call stack:
    line 140 of static/components/requirejs/require.js in defaultOnError
    line 544 of static/components/requirejs/require.js in onError
    line 1735 of static/components/requirejs/require.js
Timeout for http://localhost:8888/a@b/
Is the notebook server running?
FAIL "function () {
            return this.evaluate(function(guid) {
                return window[guid] !== undefined;
            }, {guid: guid});
        }" did not evaluate to something truthy in 10000ms
#    type: uncaughtError
#    file: notebook/tests/base/deprecated-require-paths.js
#    error: "function () {
            return this.evaluate(function(guid) {
                return window[guid] !== undefined;
            }, {guid: guid});
        }" did not evaluate to something truthy in 10000ms

#    stack: not provided
```

takluyver commented:
Oh yes, the mysterious 'Script error'...

@jdfreder should be able to give you some pointers on what's needed - we just landed his big PR that changes how the JS is built.

jdfreder commented on notebook/static/notebook/js/celltoolbarpresets/attachments.js:
```diff
@@ -0,0 +1,51 @@
+// Distributed under the terms of the Modified BSD License.
+
+define([
+    'jquery',
```
No need for this import, it's a global now

jdfreder replied:
And here

jdfreder commented on notebook/static/notebook/js/celltoolbarpresets/attachments.js:
```diff
@@ -0,0 +1,51 @@
+    'jquery',
+    'notebook/js/celltoolbar',
+    'base/js/dialog',
+], function($, celltoolbar, dialog) {
```
Same here with the $

jdfreder replied:
idem

jdfreder commented:
Webpack is a little more strict about dependencies, try removing the jquery imports I mention, although I'm surprised they'd cause it to fail, that's an issue I'll need to investigate...

The error you see is because `notebook/static/notebook/js/celltoolbarpresets/attachments.js` cannot load.

julienr commented:
@jdfreder Thanks for your help. Removing the jquery dependencies seem to have solved the travis failure. The "script error" I got locally was due to my build directory containing some old js code.

julienr commented:
I have a travis failure on `js/services` which I think is just some weird timeout. Can someone restart this test ?

julienr commented:
Thanks to whoever relaunched the test :) I've added the two tests (attachments and garbage collection) I wanted, I think this can be considered ready.

minrk commented:
Awesome, thanks @julienr!

stevengj commented:
I'm really excited about this!

julienr commented:
Thanks to everybody involved in this discussion and thanks for merging this !

jdfreder commented:
:tada: :cake: This is awesome!!

Carreau commented:
Ha ! Found a bug already ! But awesome !

platise commented:
Save as HTML or export to slides does not include attached images, but it prints a filename instead in jupyter 5.2.2. Is this a bug or any hint how to export a report?

On save as .tex I would suggest images are exported under some temp dir, like .ipynb_tmp to be able to generate a PDF.

takluyver commented:
The export functionality is part of nbconvert, and I'm not sure that it supports 'attached' images yet. See https://github.com/jupyter/nbconvert/issues/699 .

platise commented:
Looks like not many people are using attached images in markdown, how to trigger more interest to the jupyter/nbconvert#699, any idea?

platise commented:
If you merge two cells where each has one image inserted as attachment then the first (above) cell will loose the image.

stevengj commented:
@platise, that sounds like a separate issue that should be filed and linked to this one.

platise commented:
Please @stevengj if you can advise or /copy/ the bug report to a proper place; as the starting /body/ of the ticket says about so many features, thought it could belong here.
